### PR TITLE
SURT multi-talker ASR recipe

### PR DESCRIPTION
**Pull request by desh2608**
This is a multi-talker ASR recipe using the SURT model: https://arxiv.org/abs/2109.08555. Here is a figure providing an overview of the model:

surt

See also: https://github.com/lhotse-speech/lhotse/pull/951

#### IHM-Mix

| Model | # params | 0L | 0S | OV10 | OV20 | OV30 | OV40 | Avg. |
|------------|:-------:|:----:|:---:|----:|:----:|:----:|:----:|:----:|
| dprnn_zipformer (base)  | 26.7 | 5.1 | 4.2 | 13.7 | 18.7 | 20.5 | 20.6 | 13.8 |
| dprnn_zipformer (large) | 37.9 | 4.6 | 3.8 | 12.7 | 14.3 | 16.7 | 21.2 | 12.2 |

#### SDM

| Model | # params | 0L | 0S | OV10 | OV20 | OV30 | OV40 | Avg. |
|------------|:-------:|:----:|:---:|----:|:----:|:----:|:----:|:----:|
| dprnn_zipformer (base)  | 26.7 | 6.8 | 7.2 | 21.4 | 24.5 | 28.6 | 31.2 | 20.0 |
| dprnn_zipformer (large) | 37.9 | 6.4 | 6.9 | 17.9 | 19.7 | 25.2 | 25.5 | 16.9 |

## Pre-trained models and logs

* Pre-trained models: <https://huggingface.co/desh2608/icefall-surt-libricss-dprnn-zipformer>

* Training logs:
    - surt_base: <https://tensorboard.dev/experiment/YLGJTkBETb2aqDQ61jbxvQ/>
    - surt_base_adapt: <https://tensorboard.dev/experiment/pjXMFVL9RMej85rMHyd0EQ/>
    - surt_large: <https://tensorboard.dev/experiment/82HvYqfrSOKZ4w8Jod2QMw/>
    - surt_large_adapt: <https://tensorboard.dev/experiment/5oIdEgRqS9Wb6yVuxaExEw/>

### CODE REVIEW

File: egs/libricss/SURT/dprnn_zipformer/beam_search.py [hunk 1]
```diff
@@ -0,0 +1,1280 @@
+# Copyright    2021  Xiaomi Corp.        (authors: Fangjun Kuang
```
[the remark this answers is not in the document]
Comment (desh2608) [reply egs/libricss/SURT/dprnn_zipformer/beam_search.py:1]: Yeah I already replaced the ones that could be replaced. The others differ slightly from the original files. For example, in the `beam_search.py`, I use `SURT` model instead of `Transducer` with corresponding changes in the decoding methods.

[the remark this answers is not in the document]
Comment (csukuangfj) [reply egs/libricss/SURT/dprnn_zipformer/beam_search.py:1]: >  in the beam_search.py, I use SURT model instead of Transducer with corresponding changes in the decoding methods.

In the latest master, we are using `nn.Module` to replace `Transducer`, so it should be fine to replace it with a symlink.

[the remark this answers is not in the document]
Comment (desh2608) [reply egs/libricss/SURT/dprnn_zipformer/beam_search.py:1]: Ahh, I see. Thanks for pointing out.

I have replaced some other scripts with symlinks. For the particular case of beam_search.py, there are some small changes in the implementations, and I also plan to add SURT-specific decoding methods in the near future, so I think it may be okay to keep it separate for now.

[the remark this answers is not in the document]
Comment (desh2608) [reply egs/libricss/SURT/dprnn_zipformer/beam_search.py:1]: @csukuangfj do you think we can merge this soon?

**Comment by desh2608** [comment]
All tests are passing now.

**Comment by csukuangfj** [comment]
> All tests are passing now.

Sorry, missed it again.

File: egs/libricss/SURT/dprnn_zipformer/decode.py [hunk 2]
```diff
@@ -0,0 +1,653 @@
...
+    feature = feature.to(device)
+    feature_lens = batch["input_lens"].to(device)
+
+    # Apply the mask encoder
+    B, T, F = feature.shape
+    processed = model.mask_encoder(feature)  # B,T,F*num_channels
+    masks = processed.view(B, T, F, params.num_channels).unbind(dim=-1)
+    x_masked = [feature * m for m in masks]
+
+    # To save the masks, we split them by batch and trim each mask to the length of
```
Comment (csukuangfj) [inline egs/libricss/SURT/dprnn_zipformer/decode.py:236]: When
https://github.com/k2-fsa/icefall/blob/75e5c2775dc5f750d23c30c76d5153a26cd90b30/egs/libricss/SURT/dprnn_zipformer/decode.py#L181
is False, can we skip this step?

File: egs/libricss/SURT/README.md [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+The model is based on the papers:
+
+- Lu, Liang et al. “Streaming End-to-End Multi-Talker Speech Recognition.” IEEE Signal Processing Letters 28 (2020): 803-807.
+- Raj, Desh et al. “Continuous Streaming Multi-Talker ASR with Dual-Path Transducers.” ICASSP 2022 - 2022 IEEE International Conference on Acoustics, Speech and Signal Processing (ICASSP) (2021): 7317-7321.
+
+The model is a combination of a speech separation model and a speech recognition model,
+but trained end-to-end with a single loss function. The overall architecture is shown
+in the figure below. Note that this architecture is slightly different from the one
+in the above papers. A detailed description of the model can be found in the following
+paper: [SURT 2.0: Advanced in transducer-based multi-talker ASR]().
```
[the remark this answers is not in the document]
Comment (desh2608) [reply egs/libricss/SURT/README.md:44]: Done, thanks!

File: egs/libricss/SURT/README.md [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+    --num-mask-encoder-layers 4 \
+    --num-encoder-layers 2,2,2,2,2
+```
+
+The above is for SURT-base (~26M). For SURT-large (~38M), use:
+
+```bash
+    --num-mask-encoder-layers 6 \
+    --num-encoder-layers 2,4,3,2,4 \
+    --model-init-ckpt exp/zipformer_large.pt \
```
[the remark this answers is not in the document]
Comment (desh2608) [reply egs/libricss/SURT/README.md:153]: See above section on "Prerequisites". The `zipformer_base.pt` and `zipformer_large.pt` are already uploaded in the HF repo.

File: egs/libricss/SURT/README.md [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+    --decoder-dim 512 \
+    --joiner-dim 512
+```
+
+The above is for SURT-base (~26M). For SURT-large (~38M), use `--num-encoder-layers 2,4,3,2,4`.
+
+Once the above model is trained for 10 epochs, copy it to `egs/libricss/SURT/exp`:
+
+```bash
+cp -r pruned_transducer_stateless7_streaming/exp/epoch-10.pt exp/zipformer_base.pt
```
[the remark this answers is not in the document]
Comment (desh2608) [reply egs/libricss/SURT/README.md:115]: > Could you also upload
> 
> * `pruned_transducer_stateless7_streaming/exp/epoch-10.pt`
Same as zipformer_base.pt which is already uploaded.
> * `dprnn_zipformer/exp/surt_base/epoch-30.pt`
Done
> * `data/lang_bpe_500/{tokens.txt, words.txt}`
Done
> * `dprnn_zipformer/exp/surt_base_adapt/epoch-8.pt`
Done
> * `dprnn_zipformer/exp/surt_large/epoch-30.pt`
Done
> * `dprnn_zipformer/exp/surt_large_adapt/epoch-8.pt`
Done
> 
> Uploading the final `epoch-xx.pt` makes it easier for users to either continue the training or fine-tune the provided models.

For the "adapt" models, the final checkpoint is same as the pretrained.pt since we don't use model averaging for them. This is because the adaptation data is very small (1 hour) and averaging models was resulting in WER degradation.

File: egs/libricss/SURT/README.md [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+
+To decode the model, run the following from within `egs/libricss/SURT`:
+
+#### Greedy search
+
+```bash
+export CUDA_VISIBLE_DEVICES="0"
+
+python dprnn_zipformer/decode.py \
+    --epoch 8 --avg 1 --use-averaged-model False \
```
[the remark this answers is not in the document]
Comment (desh2608) [reply egs/libricss/SURT/README.md:208]: Since we use the dev data for adaptation, we cannot really do a grid search on these options since the WER on dev is quite small (since we train on it), and using test WER for the search is cheating. As such, I did not do an exhaustive search but rather used the train and valid loss to stop early.

File: egs/libricss/SURT/dprnn_zipformer/decode.py (hunk 2 above)
Comment (desh2608) [reply egs/libricss/SURT/dprnn_zipformer/decode.py:236]: Yes. Added.

File: egs/libricss/SURT/README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (csukuangfj) [reply egs/libricss/SURT/README.md:153]: Could you point me to the exact location? I could not find them in the HF repo.

[the remark this answers is not in the document]
Comment (desh2608) [reply egs/libricss/SURT/README.md:153]: See: https://huggingface.co/desh2608/icefall-surt-libricss-dprnn-zipformer/tree/main/exp

**Comment by desh2608** [comment]
FYI, the paper describing this recipe is available on ArXiv: https://arxiv.org/abs/2306.10559 :)

**Comment by csukuangfj** [comment]
@yaozengwei  is helping review it.

**Comment by desh2608** [comment]
> @yaozengwei  is helping review it.

Thanks!

**Review by yaozengwei — APPROVED** [review]
